snooty/parser.py

```diff
@@ -419,7 +430,7 @@ def dispatch_visit(self, node: docutils.nodes.Node) -> None:
             raise docutils.nodes.SkipNode()
         elif isinstance(node, rstparser.snooty_diagnostic):
             self.diagnostics.append(node["diagnostic"])
-            return
+            raise docutils.nodes.SkipNode()
```

Comment:
Just for my understanding: are we using `SkipNode` instead of `return` here because of the way diagnostics nodes would cause wonky behavior with removing lists but not list items?

Reply:
So the way the recursion works is that `dispatch_visit()` creates a node to append to the state stack, and `dispatch_departure()` then pulls that node off and adds it to its parent's `children` list. Just returning here doesn't create a node, but still calls `dispatch_departure()` to pop the most recent node. SkipNode prevents calling `dispatch_departure()`.

This whole class is a hot mess, but rewriting it is out of scope right now :D

Reply:
Thank you for this explanation!